gst/nnstreamer/tensor_sparse/tensor_sparse_enc.c

```diff
@@ -0,0 +1,420 @@
+
+  if (filter) {
+    GstCaps *intersection;
+    intersection =
+        gst_caps_intersect_full (filter, caps, GST_CAPS_INTERSECT_FIRST);
+
+    gst_caps_unref (caps);
+    caps = intersection;
+  }
+
+  silent_debug_caps (self, caps, "result");
+  return caps;
+}
+
+/**
+ * @brief This function handles sink pad query.
+ */
+static gboolean
+gst_tensor_sparse_enc_sink_query (GstPad * pad, GstObject * parent,
+    GstQuery * query)
+{
+  GstTensorSparseEnc *self;
+
+  self = GST_TENSOR_SPARSE_ENC (parent);
+
+  GST_DEBUG_OBJECT (self, "Received %s query: %" GST_PTR_FORMAT,
+      GST_QUERY_TYPE_NAME (query), query);
+
+  switch (GST_QUERY_TYPE (query)) {
+    case GST_QUERY_CAPS:
+    {
+      GstCaps *caps;
+      GstCaps *filter;
+
+      gst_query_parse_caps (query, &filter);
+      caps = gst_tensor_sparse_enc_query_caps (self, pad, filter);
+      silent_debug_caps (self, filter, "filter");
+      gst_query_set_caps_result (query, caps);
+      gst_caps_unref (caps);
+      return TRUE;
+    }
+    case GST_QUERY_ACCEPT_CAPS:
+    {
+      GstCaps *caps;
+      GstCaps *template_caps;
+      gboolean res = FALSE;
+
+      gst_query_parse_accept_caps (query, &caps);
+      silent_debug_caps (self, caps, "caps");
+
+      if (gst_caps_is_fixed (caps)) {
+        template_caps = gst_pad_get_pad_template_caps (pad);
+
+        res = gst_caps_can_intersect (template_caps, caps);
+        gst_caps_unref (template_caps);
+      }
+
+      gst_query_set_accept_caps_result (query, res);
+      return TRUE;
+    }
+    default:
+      break;
+  }
+
+  return gst_pad_query_default (pad, parent, query);
+}
+
+
+/**
+ * @brief Internal function to transform the input buffer.
+ */
+static GstFlowReturn
+gst_tensor_sparse_enc_chain (GstPad * pad, GstObject * parent, GstBuffer * buf)
+{
+  GstFlowReturn ret = 0;
+  GstTensorSparseEnc *self = GST_TENSOR_SPARSE_ENC (parent);
+  GstMemory *mem, *out_mem;
+  GstMapInfo map;
+  GstBuffer *outbuf;
+  GstTensorsInfo *info;
+  guint i;
+
+  UNUSED (pad);
+
+  info = &self->in_config.info;
+
+  outbuf = gst_buffer_new ();
+
+  for (i = 0; i < info->num_tensors; ++i) {
+    GstTensorMetaInfo meta;
+    gst_tensor_info_convert_to_meta (&info->info[i], &meta);
+
+    meta.format = _NNS_TENSOR_FORMAT_SPARSE;
+    meta.media_type = _NNS_TENSOR;
+
+    /* do real encoding here */
+    mem = gst_buffer_peek_memory (buf, i);
+    if (!gst_memory_map (mem, &map, GST_MAP_READ)) {
+      ml_loge ("Cannot map input buffer to gst-buffer at sparse_enc.");
+      gst_memory_unref (mem);
+      gst_buffer_unref (outbuf);
+      return GST_FLOW_ERROR;
+    }
+
+    out_mem = gst_tensor_sparse_from_dense (&meta, map.data);
+    gst_memory_unmap (mem, &map);
+
+    if (!out_mem) {
+      nns_loge ("failed to convert to sparse tensor");
+      gst_buffer_unref (outbuf);
+      return GST_FLOW_ERROR;
+    }
+
+    gst_buffer_append_memory (outbuf, out_mem);
+  }
+
+  ret = gst_pad_push (self->srcpad, outbuf);
+
```

Comment:
need to un-reference the `buf` in an error or normal finish.
 ```suggestion
 done:
  gst_buffer_unref (buf);
  return ret;
```

Reply:
Thanks for the suggestion! I updated the code after this. Please check.